c2rust-analyze/src/util.rs

```diff
@@ -145,11 +146,13 @@ pub fn ty_callee<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Callee<'tcx>>
     })
 }
 
-fn builtin_callee<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    did: DefId,
-    _substs: SubstsRef<'tcx>,
-) -> Option<Callee<'tcx>> {
+fn builtin_callee<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, did: DefId) -> Option<Callee<'tcx>> {
+    let is_trivial = ty.fn_sig(tcx).is_trivial(tcx);
```

Comment:
`ty.fn_sig(tcx)` will return sigs for `FnPtr`s and `Closure`s.  Replacing it with `tcx.fn_sig(did)` I think makes it clearer that we only expect to be working with `FnDef`s here.  It would also let you revert the change to the signature of `builtin_callee`.

Reply:
`ty.fn_sig`, 


```rust
    pub fn fn_sig(self, tcx: TyCtxt<'tcx>) -> PolyFnSig<'tcx> {
        match self.kind() {
            FnDef(def_id, substs) => tcx.bound_fn_sig(*def_id).subst(tcx, substs),
            FnPtr(f) => *f,
            Error(_) => {
                // ignore errors (#54954)
                ty::Binder::dummy(FnSig::fake())
            }
            Closure(..) => bug!(
                "to get the signature of a closure, use `substs.as_closure().sig()` not `fn_sig()`",
            ),
            _ => bug!("Ty::fn_sig() called on non-fn type: {:?}", self),
        }
    }
```

doesn't return anything for `Closure`s, only for `FnDef` and `FnPtr`.  For `FnPtr`, I don't see why we'd treat function pointer calls different from function calls, as there's no hidden closure, and either way we just calculate the triviality based on the argument and return types.  `builtin_callee` is only called after we previously checked for `FnDef`, but we can relax that check when the need arises.  And if I use `tcx.fn_sig` instead, then I have to needlessly repeat the whole `tcx.bound_fn_sig(*def_id).subst(tcx, substs)`.